kotlinx-coroutines-core/common/src/CoroutineStart.kt

```diff
@@ -1,74 +1,343 @@
+     * runBlocking {
+     *     println("1. About to start a coroutine not needing a dispatch.")
+     *     // Dispatch the job to execute.
+     *     // `Dispatchers.Unconfined` is explicitly chosen.
+     *     val job = launch(Dispatchers.Unconfined) { // CoroutineStart.DEFAULT is the launch's default start mode
+     *         println("2. The body will be executed immediately")
+     *         delay(50.milliseconds) // give up the thread to the outer coroutine
+     *         println("4. When the thread is next available, this coroutine proceeds further")
+     *     }
+     *     println("3. After the initial suspension, the thread does other work.")
+     * }
+     * ```
+     *
+     * ```
+     * // Example of cancelling coroutines before they start executing.
+     * runBlocking {
+     *     // dispatch the job to execute on this thread later
+     *     launch { // CoroutineStart.DEFAULT is the launch's default start mode
+     *         println("This code will never execute")
+     *     }
+     *     cancel() // cancels the current coroutine scope and its children
+     *     launch(Dispatchers.Unconfined) {
+     *         println("This code will never execute")
+     *     }
+     *     println("This code will execute.")
+     * }
+     * ```
      */
     DEFAULT,
 
     /**
      * Starts the coroutine lazily, only when it is needed.
      *
-     * See the documentation for the corresponding coroutine builders for details
-     * (like [launch][CoroutineScope.launch] and [async][CoroutineScope.async]).
+     * Starting a coroutine with [LAZY] only creates the coroutine, but does not schedule it for execution.
+     * When the completion of the coroutine is first awaited
+     * (for example, via [Job.join]) or explicitly [started][Job.start],
+     * the dispatch procedure described in [DEFAULT] happens in the thread that did it.
+     *
+     * The details of what counts as waiting can be found in the documentation of the corresponding coroutine builders
+     * like [launch][CoroutineScope.launch] and [async][CoroutineScope.async].
+     *
+     * If the coroutine's [Job] is cancelled before it started executing, then it will not start its
+     * execution at all and will be considered [cancelled][Job.isCancelled].
      *
-     * If coroutine [Job] is cancelled before it even had a chance to start executing, then it will not start its
-     * execution at all, but will complete with an exception.
+     * **Pitfall**: launching a coroutine with [LAZY] without awaiting or cancelling it at any point means that it will
+     * never be completed, leading to deadlocks and resource leaks.
+     * For example, the following code will deadlock, since [coroutineScope] waits for all of its child coroutines to
+     * complete:
+     * ```
+     * coroutineScope {
+     *     launch(start = CoroutineStart.LAZY) { }
+     * }
+     * ```
+     *
+     * Behavior of [LAZY] can be described with the following examples:
```

Comment:
I probably understand what `@globsterg` means when they say they didn't find a use case for `LAZY`. *How* does one use `LAZY`? Typically, in a pattern like
```kotlin
// a property
val optionallyRequiredValue = scope.async(start = CoroutineStart.LAZY) {
    // ...
}
```
But I don't see how this is better than
```kotlin
val optionallyRequiredValue by lazy {
    scope.async {
        // ...
    }
}
```
The latter option doesn't spawn unnecessary new child coroutines (which could stop the scope from completing if you aren't careful).

Or let's say you have something similar to
```kotlin
val sendOnes = scope.launch(start = CoroutineStart.LAZY) {
    while (true) { onesChannel.send(1) }
}
```
Then, you have to write things like `sendOnes.start(); channel.receive()` throughout your codebase. But how is this better than something like the following?
```kotlin
val onesChannel by lazy {
    Channel<Int>().also {
        scope.launch {
            while (true) { onesChannel.send(1) }
        }
    }
}
```

I can sort of see the utility of `LAZY` when you have to combine on-demand computation with the ability to reinitialize the variable anew:

```kotlin
@Volatile // or @Synchronized
var latestResult: Deferred<Int> = computeResult()

private fun computeResult(): Deferred<Int> = scope.async(start = CoroutineStart.LAZY) {
    // ...
}

fun update() {
    // latestResult?.let { it.cancel() } // can't do that: the clients will throw
    latestResult = computeResult()
}
```

But this looks to me like a good fit for a hot `Flow` chain: mutable state in concurrent scenarios is tricky to get right, it's best avoided, not to mention that this allows for many parallel `computeResult()` computations, which can DDoS the system.

I'm really struggling to think of a use case where `CoroutineStart.LAZY` is actually the superior option. When I search grep.app for `LAZY` usages, most of them are completely unclear to me and seem to have been created by accident, like https://github.com/Baeldung/kotlin-tutorials/blob/f203e8fd8571b8dcc313fca04c72196bc4948649/spring-boot-kotlin/src/main/kotlin/com/baeldung/nonblockingcoroutines/controller/ProductControllerCoroutines.kt#L31-L44 Why is this `LAZY` if it's immediately started?

https://github.com/JetBrains/compose-multiplatform/blob/6aad20ec087f54df168ad7e7ce03c4ded710a93c/components/resources/library/src/commonMain/kotlin/org/jetbrains/compose/resources/AsyncCache.kt could be a mistake, or it could actually be dictated by some requirement to avoid deadlocks if `CoroutineDispatcher.dispatch` calls `getOrLoad`. I'm not sure.

Do you maybe know of some (semi-)real-world examples where `CoroutineStart.LAZY` is a good choice?

Reply:
Additionally: It seems like we should somehow include this information about CoroutineStart.LAZY and how to replace it with with `by lazy` in the documentation as well: Otherwise, simply by looking at the shape of the API, it's easy for people to surmise that `CoroutineStart.LAZY` should just be used for all the aforementioned cases. I don't know if you have a preference on whether it should be an inlined part to the docstring, or if it would make more sense to expand the coroutines guide and link it? It'd be great to avoid requiring users to make this replacement discovery by themselves.

Reply:
> A cursory search yielded https://cs.android.com/android-studio/platform/tools/adt/idea/+/mirror-goog-studio-main:preview-designer/src/com/android/tools/idea/preview/PreviewRefreshManager.kt?q=CoroutineStart.LAZY&start=41

I think I understand what's going on there: the coroutine needs to access a value that's otherwise only available under the lock. In essence,
```kotlin
val coroutine: Job // local variable
lock.withLock {
  val value = obtain()
  currentValue = value // currentValue is protected by the lock
  coroutine = launch(start = CoroutineStart.LAZY) {
    doWork(value)
  }
}
coroutine.invokeOnCompletion { cleanup() }
coroutine.join()
```
Doesn't seem to me any better than
```kotlin
val value: Job // local variable
lock.withLock {
  value = obtain()
  currentValue = value // currentValue is protected by the lock
}
try {
  doWork(value)
} finally {
  cleanup()
}
```
In fact, their `invokeOnCompletion` lambda scares me with its complexity. Given that the lambda is supposed to never throw any exceptions and finish quickly, letting it grow so big is suspicious. So, `try`/`finally` wins even just on this account!

> It seems like we should somehow include this information about CoroutineStart.LAZY and how to replace it with with by lazy in the documentation as well

I agree.

> I don't know if you have a preference on whether it should be an inlined part to the docstring, or if it would make more sense to expand the coroutines guide and link it?

Not sure. On the one hand, this may be a bit too extensive for the KDoc, but on the other hand, our coroutine guide is completely obsolete, and I'd prefer not to touch it if I can help it until it's completely reworked. Maybe the idea that `LAZY` isn't that useful can be phrased clearly and succinctly enough to fit the KDoc. I'll give it a try.

Reply:
Great explanation, I think this really covers it.